### PR TITLE
[5.2] Build occurrences index for Merlin

### PR DESCRIPTION
**Pull request by voodoos**
This PR adds new rules that drive the `ocaml-index` tool to aggregate a workspace index. The rules are a bit simpler as the ones in https://github.com/ocaml/dune/pull/6762 since now part of the indexation will is done by the compiler in 5.2.

Building the index involves three steps:
- Use the `-bin-annot-occurrences` flag to have the compiler store partial indexes in the `cmt` files.
- Run the `ocaml-index` tool to make an index for each stanza (actually each compilation-context).
- Run the `ocaml-index` tool to aggregate every stanza's index into a global project index.
  
I ran simple benchmarks on Dune itself. These are cold builds ran after `./dune.exe clean`:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./dune.exe build @check` (without `-bin-annot-occurrences`) | 8.632 ± 0.167 | 8.454 | 8.950 | 1.00 |
| `./dune.exe build @check` | 8.637 ± 0.108 | 8.433 | 8.796 | 1.00 |
| `./dune.exe build @check @ocaml-index` | 13.597 ± 0.173 | 13.384 | 13.904 | 1.00 |

- There are no perceptible performance cost of using the `-bin-annon-occurrences` flag when writing `cmt` files, so the best option is probably to always use that flag to simplify the rules.
- Actually building and gathering the index is costly (especially from a cold build), so it should be opt-in. (We have several optimizations in mind that should improve that situation in the future.)

There are several moving parts that I would like to discuss in the next dune-dev meeting:
1. The indexing rules have specific dependencies that I am unsure how best to check/enforce:
   - The external `ocaml-index` tool must be installed
   - The `-bin-annot-occurrences` flag requires OCaml 5.2
2. I am unsure of the best way to make indexation opt-in (it can be expensive on massive projects)
   - I initially added a `dune-workspace`-level option  that I remove in cd34ae18a860a3e7aec6b6fe0f798d891cb22617 following a discussion with @emillon 
   - I currently think we could simply keep the `@ocaml-index` alias but remove it from `@default` so that users have to explicitly use it.

Other issues to consider:
- The Melange compiler is not yet accepting the `-bin-annot-occurrences` flag

image

### CODE REVIEW

File: Makefile
Comment (emillon) [inline Makefile:29]: it's best if `ocaml-index` is not a test-dep. you can write a fake binary that still allows you to test the behavior (I can help with that)

File: otherlibs/stdune/src/stdune.ml
Comment (emillon) [inline otherlibs/stdune/src/stdune.ml:2]: just so we don't forget: you can work around that by passing `-alert -unstable`

File: src/dune_rules/alias0.ml
Comment (emillon) [inline src/dune_rules/alias0.ml:17]: if we end up making a new standard index, it'll need to be documented in `doc/reference/aliases.rst`

**Comment by voodoos** [comment]
> For a first release, I would consider the following strategy to select whether this feature is enabled:
> 
>     * attach these rules to `@ocaml-index`
> 
>     * if `ocaml-index` is available, and OCaml >= 5.2 (which I suppose is an opam constraint already), attach `@ocaml-index` to `@check`
> 
> 
> (This is similar to how we set up `sherlodoc`, though the situation is a bit different)
> 
> That way:
> 
>     * you can test this in dune's test suite, by providing a fake `ocaml-index` binary (check how sherlodoc is tested)
> 
>     * people can use it in an opt-in way in their workspaces
> 
>     * we can find later a way to configure this in a more precise way.

I wonder whether we should attach `@ocaml-index` to `@check` to make it really "opt-in" ?

**Comment by voodoos** [comment]
@rgrinberg removing the aggregation step does have an impact on cold build time, which is probably more noticeable on rebuilds. `@check @ocaml-index` is still 46% slower than `@check`, but it was 57% before. 

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./dune.exe build @check` | 9.090 ± 0.089 | 8.959 | 9.222 | 1.00 |
| `./dune.exe build @check @ocaml-index` | 13.302 ± 0.174 | 12.964 | 13.449 | 1.00 |

But now I realize that my measurement are wrong: the `@ocaml-index` appears to build much more targets (and not only indexes) than `@check` so it is not a good comparison. (and maybe I do something wrong in my rules ?)

If I compare with `@check @install` the results are much "better", only a 16% increase in built time:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./dune.exe build @check @install` | 18.552 ± 0.441 | 18.206 | 19.745 | 1.00 |
| `./dune.exe build @check @install @ocaml-index` | 21.517 ± 0.219 | 21.224 | 21.976 | 1.00 |

But even then, there is some additional building happening (in vendored dirs notably). I find this a bit surprising since the dependencies of `@ocaml-index` are "every cmt of every stanza of the project" which I would expect to be the result of building `@check` ?

It would be great if someone could have a look :-)
Additionally, is there some way to list the targets built for some given aliases ? And to have Dune output time spent running commands (ideally total time per tool) ?

In any case, the more I think about it the more I am convinced that the "don't merge" strategy is the best. I will update the merlin config and see how it goes.

File: otherlibs/stdune/src/stdune.ml
Comment (rgrinberg) [inline otherlibs/stdune/src/stdune.ml:1]: an accident?

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:9]: Make this argument required please.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:51]: `additional_libs`

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:68]: Maybe you can can use `Lib_flags.L.include_flags` here?

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:77]: This should be `Command.run_dyn_prog` and the resolution of the binary should be delayed into action builder

**Comment by rgrinberg** [comment]
As discussed, the performance penalty currently is too large to build this by default. A separate alias sounds like a good start to me. When this feature becomes fast enough for general use, do you think we will still need the ocaml-index alias?

Also, could you comment on the stability of this feature for the users? When this feature is shipped, is it recommended that everyone start using it right away? Or is it more for wider experimentation?

**Comment by voodoos** [comment]
Thanks for your review @rgrinberg. I was focused on experimenting with performance improvement but I will answer your comments soon.

We spent some time with @emillon to make the indexer faster and we did found some important optimizations.

However indexing only the libraries/executables which are part of `@check` still takes  increases cold-build time by 18% .
Moreover most of the additional time is spent at the end of the build when indexing the `dune_rules` library which features a large number of modules. I notably tried things such as indexing each `cmt` independently to get finer grained dependencies. This improves partial re-build time a lot but also increase significantly cold-build time because it doesn't allow the indexer to cache results that are the same for every `cmt`.

In short, I don't think we can afford to enable the feature by default right now. I propose to add two additional targets: 
- `@check-index` which only indexes cmt generated by the `@check` target
- `@index-all` which perform full indexation of the project, which might trigger the build of additionnal targets like unused vendored dependencies.

Eventually, our best option is probably to have LSP ask dune via RPC to update the index when needed.

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./dune.exe build @check` | 8.789 ± 0.094 | 8.673 | 8.920 | 1.00 |
| `./dune.exe build @check-index` | 10.422 ± 0.174 | 10.227 | 10.838 | 1.00 |
| `./dune.exe build @index-all` | 11.520 ± 0.239 | 11.177 | 11.877 | 1.00 |

Results are actually better on a codebase such as Irmin where there is no single very large library such as `dune_rules`: the time increase between `@check` ad `@check-index`is 9.5%.

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `dune build @check` | 15.718 ± 0.091 | 15.618 | 15.928 | 1.00 |
| `dune build @check-index` | 17.220 ± 0.129 | 17.000 | 17.420 | 1.00 |
| `dune build @ocaml-index` | 17.437 ± 0.120 | 17.228 | 17.601 | 1.00 |

There is a tension between: more granularity for smaller incremental rebuild VS less granularity that allow for greater use of cache which speeds up cold builds but slow down incremental rebuilds. We could probably adapt the indexer to cache on the disk results for each cmt so that we could get the best of both worlds, but I don't think the benefit would be big enough to make it a priority.

File: Makefile
Comment (voodoos) [reply Makefile:29]: Could you point me to such "fake binary" in the tree that I could take inspiration from ?

File: otherlibs/stdune/src/stdune.ml
Comment (voodoos) [reply otherlibs/stdune/src/stdune.ml:1]: It was the quickest hack to make my build pass with OCaml 5.2, I pushed it by mistake.

File: otherlibs/stdune/src/stdune.ml
Comment (voodoos) [reply otherlibs/stdune/src/stdune.ml:2]: I am not familiar with alerts, I tried to add `-alert` but it still fails:
```
File "otherlibs/stdune/src/stdune.ml", line 2, characters 0-29:
2 | [@@@alert "-alert -unstable"]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error (warning 47 [attribute-payload]): illegal payload for attribute 'alert'.
Ill-formed list of alert settings

File "otherlibs/stdune/src/stdune.ml", line 1, characters 4-9:
1 | [@@@alert unstable "The API of this library is not stable and may change without notice."]
        ^^^^^
Error (warning 53 [misplaced-attribute]): the "alert" attribute cannot appear in this context
```

File: src/dune_rules/alias0.ml
Comment (voodoos) [reply src/dune_rules/alias0.ml]: todo @voodoos: I will add documentation when we will have settled on the additional alias(es) we want.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml]: Right, it is not even used anymore, I removed it.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml]: Done in 92c986567cd4ca1317b96aa3e3d2672aa29ae452

Do you know if there is a better way for me to get the list of implicit transitive dependencies than doing the ugly `requires_link \ requires_compile` intersection ?

File: Makefile
Comment (emillon) [reply Makefile:29]: `test/blackbox-tests/utils/refmt.ml`

File: otherlibs/stdune/src/stdune.ml
Comment (emillon) [reply otherlibs/stdune/src/stdune.ml:2]: taking a step back: why did you make that change? in other words, what breaks when you leave these attributes?

File: Makefile
Comment (voodoos) [reply Makefile:29]: I tried to do something similar but I had to hack some workaround because the binary doesn't have a valid module name (`ocaml-index`). Is there a better way to do it ?

It's in that commit: https://github.com/ocaml/dune/pull/10422/commits/cc7c74e29f61480b5d93cb782b8782027c49bd77

**Comment by rgrinberg** [comment]
> Additionally, is there some way to list the targets built for some given aliases ? And to have Dune output time spent running commands (ideally total time per tool) ?

Yes, there's the `--trace-file` flag

**Comment by rgrinberg** [comment]
Could you stick to `check-index` or `index-all` for now? The only reason vendored sources aren't included in `@check` is because of a bug.

File: src/dune_rules/alias0.ml
Comment (voodoos) [reply src/dune_rules/alias0.ml:17]: done in [4d13b4d](https://github.com/ocaml/dune/pull/10422/commits/4d13b4d1d00e3f49166bddc0e5a94d52c68a1044)

**Comment by voodoos** [comment]
I rebased the PR and made a few more changes like adding documentation and removing `@check-index`.

I think the PR is now ready for another round of review. 

@rgrinberg I still wonder about the hidden transitive deps issue. I think now that the compiler supports the `-H` flag and Merlin has corresponding configuration directives we should pass it the hidden deps using these directives Right now to get the set of hidden deps I use `requires_link \ requires_compile` and it appears to work, but is that correct ? Is there a better way to get these ?

File: src/dune_rules/merlin/merlin.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/merlin.ml:247]: What would happen to older versions of merlin if they'd encounter this directive?

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:41]: Is the peek here intentional? In case we can't resolve one of the libs, we just pretend all of them don't exist without letting the user know?

File: src/dune_rules/merlin/merlin.ml
Comment (voodoos) [reply src/dune_rules/merlin/merlin.ml:247]: Since Merlin 4.6 unknown configuration tags are ignored when `dune ocaml-merlin` is used as a configuration reader, so it should not trigger any error.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml]: That's a good catch, I removed the peek in [4c2aee3](https://github.com/ocaml/dune/pull/10422/commits/4c2aee34457b9c4e2ec78dc7f405e748a5558add)

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:60]: Why do you not run ocaml-index from `dir`? We only run compilation commands from context_dir for good error messages. Does ocaml-index output error messages that refer to source code?

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [inline src/dune_rules/merlin/ocaml_index.ml:64]: Why is it needed to absolutize the path here?

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml:60]: `ocaml-index` relies on the load-path stored in the `cmt` files. These paths are relative to the directory in which the compiler was called.

If you would prefer the indexer to run in `dir` we can have dune pass the complete load-path via the CLI.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml:64]: The indexer aggregate parsetree locations. These locations contain a path that is relative to the call-site of the compiler. When reading the index, Merlin needs a way to find the correct source file of each occurrences. The current way of doing that is that the indexer make the relative path of location absolute using the `--root` argument. 

It was the most straightforward approach, but if you'd prefer we can also keep relative paths in the locations and either store the source root path into the index or pass it to Merlin via configuration.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [reply src/dune_rules/merlin/ocaml_index.ml:60]: No, that's fine to leave it this way. It makes sense to leave a comment about this though.

File: src/dune_rules/merlin/ocaml_index.ml
Comment (rgrinberg) [reply src/dune_rules/merlin/ocaml_index.ml:64]: > When reading the index, Merlin needs a way to find the correct source file of each occurrences. The current way of doing that is that the indexer make the relative path of location absolute using the --root argument.

Shouldn't the source directory be communicated through the .merlin file? Or the equivalent configuration that we have via the `dune ocaml-merlin` command.

> It was the most straightforward approach, but if you'd prefer we can also keep relative paths in the locations and either store the source root path into the index or pass it to Merlin via configuration.

With absolute paths, we're making the rule non relocatable and unusable with the dune cache. So I would say that using a relative path is the minimum. But see my first question. I'd like to keep relying on the configuration command that merlin runs to communicate the source directory.

**Review by rgrinberg — APPROVED** [review]
LGTM.

Please add a CHANGES entry before merging

**Pull request merged**
